core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/NativeSailStore.java

```diff
@@ -435,24 +428,14 @@ public void deprecate(Resource subj, IRI pred, Value obj, Resource ctx)
 		private synchronized void acquireExclusiveTransactionLock()
 			throws SailException
 		{
-			boolean nextTxnLockAcquired = txnLockAcquired;
-			if (!nextTxnLockAcquired) {
-				txnLockManager.lock();
+			if (!txnLockManager.isHeldByCurrentThread()) {
 				try {
-					if (txnLockManager.getHoldCount() == 1) {
-						// first object
-						tripleStore.startTransaction();
-					}
-					nextTxnLockAcquired = txnLockAcquired = true;
+					txnLockManager.lock();
+					tripleStore.startTransaction();
 				}
 				catch (IOException e) {
 					throw new SailException(e);
 				}
-				finally {
```

Comment:
The lock needs to be released if any exception occurs. Simplest way of doing this is a local boolean (see the "allGood" booleans introduced in #293) that are switched to true after tripleStore.startTransaction returns successfully within the try, to stop the lock being immediately released here. That way we don't need to catch all exceptions, but we can safely reset the state to unlocked if any exceptions are thrown.